amy/communityroles/forms.py

```diff
@@ -0,0 +1,118 @@
+    class Meta:
+        model = CommunityRole
+        fields = (
+            "config",
+            "person",
+            "award",
+            "start",
+            "end",
+            "inactivation",
+            "membership",
+            "url",
+            "generic_relation_content_type",
+            "generic_relation_pk",
+        )
+        widgets = {
+            "config": HeavySelect2Widget(
+                data_view="api:communityroleconfig-list", attrs=SELECT2_SIDEBAR
+            ),
+            "person": ModelSelect2Widget(
+                data_view="person-lookup", attrs=SELECT2_SIDEBAR
+            ),
+            "award": ModelSelect2Widget(
+                data_view="award-lookup", attrs=SELECT2_SIDEBAR
+            ),
+            "membership": ModelSelect2Widget(
+                data_view="membership-lookup", attrs=SELECT2_SIDEBAR
+            ),
+            "generic_relation_content_type": forms.Select(attrs={"disabled": ""}),
+            "generic_relation_pk": HeavySelect2Widget(
+                data_view="generic-object-lookup", attrs=SELECT2_SIDEBAR
+            ),
+        }
+        labels = {
+            "generic_relation_content_type": "Generic relation object type",
+            "generic_relation_pk": "Generic relation object",
+        }
+
+    class Media:
+        js = ("communityrole_form.js",)
+
+    def __init__(self, *args, **kwargs):
+        form_tag = kwargs.pop("form_tag", True)
+        super().__init__(*args, **kwargs)
+        bootstrap_kwargs = {
+            "add_cancel_button": False,
+            "form_tag": form_tag,
+        }
+        self.helper = BootstrapHelper(**bootstrap_kwargs)
+
+    def clean(self) -> dict[str, Any]:
+        """Validate form according to rules set up in related Community Role
+        configuration."""
+        cleaned_data = super().clean()
+        errors: defaultdict[str, list[ValidationError]] = defaultdict(list)
+        config: CommunityRoleConfig = cleaned_data["config"]
+
+        # Award required?
+        if config.link_to_award and not cleaned_data["award"]:
+            errors["award"].append(
+                ValidationError(f"Award is required with community role {config}")
+            )
+
+        # Specific award badge required?
+        if (badge := config.award_badge_limit) and (award := cleaned_data["award"]):
+            if award.badge != badge:
+                errors["award"].append(
+                    ValidationError(
+                        f"Award badge must be {badge} for community role {config}"
+                    )
+                )
+
+        # Membership required?
+        if config.link_to_membership and not cleaned_data["membership"]:
+            errors["membership"].append(
+                ValidationError(f"Membership is required with community role {config}")
+            )
+
+        # Additional URL supported?
+        if not config.additional_url and cleaned_data["url"]:
```

Comment:
Here's where I'm not sure I implemented the requirements correctly.

Basically Community Role Config model can specify if Community Role created from that configuration can store additional URL.

If Config states that URL is not allowed, then the Community Role form field for URL will be disabled and form validation will not accept any URL there.